worker/src/RTC/RtpStreamSend.cpp

```diff
@@ -10,34 +10,161 @@ namespace RTC
 	static constexpr size_t MaxRequestedPackets{ 17 };
 	thread_local static std::vector<RTC::RtpStreamSend::StorageItem*> RetransmissionContainer(
 	  MaxRequestedPackets + 1);
+	static constexpr uint32_t MinRetransmissionDelay{ 200 };
 	// Don't retransmit packets older than this (ms).
 	static constexpr uint32_t MaxRetransmissionDelay{ 2000 };
 	static constexpr uint32_t DefaultRtt{ 100 };
+	static constexpr uint16_t MaxSeq = std::numeric_limits<uint16_t>::max();
+	static constexpr uint32_t MaxTs  = std::numeric_limits<uint32_t>::max();
 
 	static void resetStorageItem(RTC::RtpStreamSend::StorageItem* storageItem)
 	{
 		MS_TRACE();
 
 		MS_ASSERT(storageItem, "storageItem cannot be nullptr");
 
-		delete storageItem->packet;
-
-		storageItem->packet     = nullptr;
+		storageItem->clonedPacket.reset();
+		storageItem->originalPacket.reset();
 		storageItem->resentAtMs = 0;
 		storageItem->sentTimes  = 0;
 		storageItem->rtxEncoded = false;
 	}
 
+	RtpStreamSend::StorageItem* RtpStreamSend::StorageItemBuffer::GetFirst() const
+	{
+		return this->Get(this->startSeq);
+	}
+
+	RtpStreamSend::StorageItem* RtpStreamSend::StorageItemBuffer::Get(uint16_t seq) const
+	{
+		auto idx{ static_cast<uint16_t>(seq - this->startSeq) };
+
+		if (this->buffer.empty() || idx > static_cast<uint16_t>(this->buffer.size() - 1))
+			return nullptr;
+
+		return this->buffer.at(idx);
+	}
+
+	bool RtpStreamSend::StorageItemBuffer::Insert(uint16_t seq, StorageItem* storageItem)
+	{
+		if (this->buffer.empty())
+		{
+			this->startSeq = seq;
+			this->buffer.push_back(storageItem);
+
+			return true;
+		}
+
+		if (seq > this->startSeq)
+		{
+			auto idx{ static_cast<uint16_t>(seq - this->startSeq) };
+
+			// Packet arrived out of order, so we already have a slot allocated for it.
+			if (idx <= static_cast<uint16_t>(this->buffer.size() - 1))
+			{
+				MS_ASSERT(this->buffer[idx] == nullptr, "Must insert into empty slot");
+				this->buffer[idx] = storageItem;
+
+				return true;
+			}
+		}
+
+		// Calculate how many elements would it be necessary to add when pushing new item to the back of
+		// the deque.
+		auto addToBack{ static_cast<uint16_t>(seq - (this->startSeq + this->buffer.size() - 1)) };
+		// Calculate how many elements would it be necessary to add when pushing new item to the front
+		// of the deque.
+		auto addToFront{ static_cast<uint16_t>(this->startSeq - seq) };
+
+		// Select the side of deque where fewer elements need to be added, while preferring the end.
+		if (addToBack <= addToFront)
+		{
+			// Packets can arrive out of order, add blank slots.
+			for (uint16_t i{ 1 }; i < addToBack; ++i)
+				this->buffer.push_back(nullptr);
+
+			this->buffer.push_back(storageItem);
+		}
+		else
+		{
+			// Packets can arrive out of order, add blank slots.
+			for (uint16_t i{ 1 }; i < addToFront; ++i)
+				this->buffer.push_front(nullptr);
+
+			this->buffer.push_front(storageItem);
+			this->startSeq = seq;
+		}
```

Comment:
This is definitely not correct. We can't associate the `startSeq` with the new packet being inserted just without invalidating the rest.

```suggestion
		else
		{
			// Packets can arrive out of order, add blank slots.
			for (uint16_t i{ 1 }; i < addToFront; ++i)
				this->buffer.push_front(nullptr);

			this->buffer.push_front(storageItem);
			this->startSeq = seq;
		}
```

Reply:
Let's just `push_back()` always.

Reply:
https://github.com/nazar-pc/mediasoup/pull/9